File: stage3/problem23.java
import java.util.HashSet;
import java.util.Iterator;
import java.util.stream.Collectors;

public class problem23 {
    
    public static void main(String[] args) {
        System.out.println(sumOfNonAbundantNumbers(10000));
        System.out.println(sumOfNonAbundantNumbers(15000));
        System.out.println(sumOfNonAbundantNumbers(20000));
        System.out.println(sumOfNonAbundantNumbers(28123));
    }
    
    private static int sumOfNonAbundantNumbers(int n) {
        HashSet<Integer> abundants = new HashSet<Integer>(n/4);
        int sum = n*(n+1)/2;
        for (var i = 1; i <= n; i++) { // start with 12 because why not
            if (isAbundant(i)) {
                abundants.add(i);
            } 
        }
        for (var i = 1; i <= n; i++) {
            boolean found = false;
            Iterator<Integer> it = abundants.iterator();
            while (it.hasNext() && !found) {
              int abundant = it.next();  
              if (abundants.contains(i - abundant)) {
                  found = true;
                  sum -= i;
              }
            }
          }
        
        return sum;
    }
    
    private static int findDivisors(int n) {
        HashSet<Integer> divisors = new HashSet<Integer>(n/2);
        divisors.add(1);
        int remainder = n;
        for (int i = 2; i < n; i++) {
            final int copyOfI = i;
            while (remainder%i == 0) {
                divisors.addAll(divisors.stream().map(a -> a*copyOfI).collect(Collectors.toSet()));
                remainder /= i;
            }
        } // unfortunately this generates all divisors, with 1 and n. (Not even a problem in this one)
        return divisors.stream().mapToInt(i -> i).sum(); 
    }
    
    private static boolean isAbundant(int n) {
        return (2*n < findDivisors(n));
    }
    
}

File: stage3/problem23.js
function sumOfNonAbundantNumbers(n) {
  // Generate Array of all sums of abundant integers
  // return n*n+1/2 - 

  var abundants = new Set();
  var sum = n * (n+1) / 2;
  for (var i = 1; i <= n; i++) { // start with 12 because why not
    if (isAbundant(i)) {
      abundants.add(i);
    } 
  }
  for (var i = 1; i <= n; i++) {
    var found = false;
    abundants.forEach(function (abundantNumber) {
      if (!found) {
        if (abundants.has(i - abundantNumber)) {
          found = true; 
          sum -= i;
        }
      } 
    })
  }
  return sum;
}

function findDivisors(n) {
  var divisors = [1];
  var remainder = n;
  for (var i = 2; i < n; i++) {
    while (remainder%i == 0) {
      divisors = divisors.concat(divisors.map(a => a*i));
      remainder /= i;
      //console.log(divisors);
    }
  } // unfortunately this generates all divisors, with 1 and n. (Not even a problem in this one)
  return Array.from(new Set(divisors)).reduce((a,b) => a+b, 0); // convert to set and back to get rid of duplicates
}

function isAbundant(n) {
  return (2*n < findDivisors(n));
}
// No idea where this goes wrong, the corresponding Java implementation is correct and relatively fast imo
console.log(sumOfNonAbundantNumbers(10000));
